resolver.go

```diff
@@ -140,38 +142,67 @@ func (r *resolver) Start() error {
 	if r.err != nil {
 		return r.err
 	}
+	// startCh is to seralize resolver Start and Stop
+	r.Lock()
+	r.startCh = make(chan struct{})
+	r.Unlock()
 
 	if err := r.setupIPTable(); err != nil {
 		return fmt.Errorf("setting up IP table rules failed: %v", err)
 	}
 
 	s := &dns.Server{Handler: r, PacketConn: r.conn}
-	r.server = s
 	go func() {
 		s.ActivateAndServe()
 	}()
 
 	tcpServer := &dns.Server{Handler: r, Listener: r.tcpListen}
-	r.tcpServer = tcpServer
 	go func() {
 		tcpServer.ActivateAndServe()
 	}()
+
+	r.Lock()
+	r.server = s
+	r.tcpServer = tcpServer
+	close(r.startCh)
+	r.startCh = nil
+	r.Unlock()
+
 	return nil
 }
 
-func (r *resolver) Stop() {
-	if r.server != nil {
-		r.server.Shutdown()
-	}
-	if r.tcpServer != nil {
-		r.tcpServer.Shutdown()
+func (r *resolver) waitForStart() {
```

Comment:
As clarified earlier there can be more than one Start and Stop calls for a resolver. This change serializes a pair of Start & Stop calls. With the suggested change a subsequent Start can block on a previous Stop which is to be avoided.

Besides, there is a personal preference rather than technical correctness aspect here. To me having a function `waitForStart` conveys the intent more clearly and easier to follow.